### ISSUE TITLE
Miscellaneous tests to add (from legacy notebooks)

### ISSUE THREAD

**Issue by pchlenski** [pchlenski/manify#16]
# Tests to roll in from `notebooks`:
* `2_polblogs_benchmark.ipynb`:
    * Verify that ProductDT = Sklearn DT for Euclidean manifolds
    * Verify `dist_component_by_manifold` sums to 1
    * __NEW__: Check equivalence to Euclidean case for other models as well
* `3_verify_shapes.ipynb`:
    * Assert `Manifold.pdist()` is close to sqrt of `Manifold.pdist2()`
    * __NEW__: Check this for all `dist2` functions
    * Assert pdists is sum of squared dists:
        ```python
        assert torch.allclose(pdists[j, k], M.dist2(x_embed[j, dims], x_embed[k, dims]).sum(), atol=1e-6)
        ```
* `5_sampling.ipynb`:
    * Check that distances to origin are the same for all wrapped normal distributions (except spherical for very high curvature)
    * Check that log-likelihoods are generally positive (Q(z) - P(z)): (repeat with ProductManifolds)
        ```python
        for K in [-2, -1.0, -0.5, 0, 0.5, 1.0, 2.0]:
            print(K)
            m = Manifold(K, 4)
            # Pick a random point to use as the center
            mu = m.sample(m.mu0)
            Sigma = torch.diag(torch.randn(m.dim)) ** 2
            samples = m.sample(z_mean=torch.cat([mu] * N_SAMPLES, dim=0), sigma=Sigma)
            log_probs_p = m.log_likelihood(z=samples)  # Default args
            log_probs_q = m.log_likelihood(z=samples, mu=mu, sigma=Sigma)
            print(
                f"Shape: {log_probs_p.shape},\tP(z) = {log_probs_p.mean().item():.3f},\tQ(z) = {log_probs_q.mean().item():.3f},\tQ(z) - P(z) = {log_probs_q.mean().item() - log_probs_p.mean().item():.3f}"
            )
            print()
        ```
    * Check that KL divergence is equal to this difference
* `10_torchified_hyperdt.ipynb`:
    * Implement legacy (iterative) version of MCDT class
    * Assert info gains are the same
    * Assert splits are the same
    * __TODO__: revisit this more carefully. Probably notebook 13 supersedes all of this.
* `13_information_gain.ipynb`:
    * Verify comparisons tensor (`verification_tensor`)
    * Verify information gains (`ig_gains_nonan`)
    * Verify angles (`angles`)
* `14_covariance_scaling.ipynb`:
    * Verify dividing variance by dimension gives you ~same norm of spacelike dimensions
* `17_verify_new_mse.ipynb`:
    * Similar to notebook 13 - need to compare information gains in regression setting
* `47_stereographic_tests.ipynb`:
    * Does the ReLU decorator work correctly?
    * Stereographic projections are ~invertible
    * mu0 becomes 0 in stereographic projection
    * Projection matrix `man.projection_matrix` is identity matrix in stereographic manifolds
    * Check that our complicated left-multiplication is equivalent to this:
    ```python
        def left_multiply(self, A, X):
        out = torch.zeros_like(X)
        for i, (A_i, X_i) in enumerate(zip(A, X)):
            m_i = self.manifold.manifold.weighted_midpoint(xs=X_i, weights=A_i)
            out[i] = self.manifold.manifold.mobius_scalar_mul(r=A_i.sum(), x=m_i)
        return out
    ```
    * Verify stereographic logits are smooth at kappa=0 (how?)

**Comment by pchlenski** [comment]
Link to notebooks branch: https://github.com/pchlenski/manify/tree/notebook_archive/notebooks

**Comment by richarddushime** [comment]
Hi mentor 
I am very interested in this project 
and I would like to contribute  but i would really need some guidance  from you 

could you  tell me what to consider on this issue ? as a starter 

i m thinking to create a new test file under tests/test_miscellaneous.py and implement the tests there 

Recommended by [OnlyDust](https://onlydust.com/) for outstanding experience and quality contributions.

**Comment by pchlenski** [comment]
Hi Richard @richarddushime, I'd be happy to guide you through some work on this project. Let me start by providing a bit of context for what this issue is about.

This library started off as a utility for personal use - mostly because I needed tools that didn't exist yet when writing the paper on mixed-curvature decision trees. I did most of the early development for Manify in Jupyter notebooks, and gradually moved code into the `src` directory as it matured. Over time, this grew into a proper open-source library and I moved the notebooks to the `notebook_archive` branch so as not to be too distracting.

When I archived the notebooks, I went through all of them looking for any assertions/tests I wrote that hadn't already been added as unit tests. This issue, in its current state, is the list that I wrote. It would be very helpful if you could go through at least some of these notebooks and try to turn them into tests, but be warned that much of the code in the notebooks may no longer work as intended, since the underlying code may have changed. As an example, you will probably see imports from an `embedders` library, which is the previous name for Manify (before we renamed it to reflect its broadened scope).

To assist you as an outside contributor, I've taken the liberty of pulling out what I think will be the easiest tests to get started with, and adding some details on where they could be added. I've also added links to the relevant notebooks/tests where possible. I hope this is sufficient to get you started, but please feel free to reach out to me at [p@chlenski.com](mailto:p@chlenski.com) if you need. Once you complete a PR for these easy tests, I can provide you significantly more detail about the remaining tests. 

## Tests to roll in from `notebooks` (by difficulty):
### Easy
* Equivalence between `dist2`/`pdist2` and `dist`/`pdist` squared:
  * **Relevant notebooks:** [`3_verify_shapes.ipynb`](https://github.com/pchlenski/manify/blob/notebook_archive/notebooks/3_verify_shapes.ipynb)
  * This one is super simple. The idea is that `dist` gives you geodesic distances between two sets of points, whereas `dist2` gives you squared distances; similarly, `pdist` gives you pairwise distances between one set of points, whereas `pdist2` gives you pairwise squared distances. We should check that e.g. `dist2(x, y)` is close to `dist(x,y) ** 2`.
  * Shockingly, we haven't actually written this assertion yet
  * The natural place for this to go is [in `test_manifolds::_shared_tests`](https://github.com/pchlenski/manify/blob/main/tests/test_manifolds.py#L53), under "Squared Distances"
  * Make sure to check the other way around too: i.e. that the square root of dist2 is close to dist. This should be the same, of course, but the point of testing is to verify rather than assume.
* Product manifold `dist2` are sum of component manifold `dist2`:
  * **Relevant notebooks:** [`3_verify_shapes.ipynb`](https://github.com/pchlenski/manify/blob/notebook_archive/notebooks/3_verify_shapes.ipynb)
  * assert torch.allclose(pdists[j, k], M.dist2(x_embed[j, dims], x_embed[k, dims]).sum(), atol=1e-6)
  * The natural place for this to go would be [in `test_manifolds::test_product_manifold_methods](https://github.com/pchlenski/manify/blob/notebook_archive/notebooks/3_verify_shapes.ipynb), as a separate test. Use the `pm.P` attribute and `pm.factorize` to break the embeddings into components, take their distances individually, and then sum them.
* Verify `dist_component_by_manifold` sums to 1:
  * **Relevant notebooks:** [`2_polblogs_benchmark.ipynb`](https://github.com/pchlenski/manify/blob/notebook_archive/notebooks/2_polblogs_benchmark.ipynb), specifically these lines:
    ```python
    contributions = embedders.metrics.dist_component_by_manifold(pm2, pm2.x_embed)
    print(contributions)
    print(sum(contributions))
    ```
  * The [`dist_component_by_manifold`](https://github.com/pchlenski/manify/blob/notebook_archive/notebooks/3_verify_shapes.ipynb) function breaks down the variance in distances by component manifolds. When you write the above tests factorizing the distances, you may as well assert that these sum to 1 as well.

**Comment by pchlenski** [comment]
Implemented by @richarddushime in #22 - thanks for your contribution!
